File: sdks/flutter/lib/src/actions.dart
import 'dart:async';
import 'dart:convert';

import 'cta.dart';
import 'queue.dart';
import 'types.dart';

// Local mirror of the mobile manifest entry (zero deps, like the snippet's).
class _Localized {
  final String title;
  final String body;
  final CtaContent? cta;
  const _Localized({required this.title, required this.body, this.cta});

  static _Localized fromJson(Map j) {
    CtaContent? cta;
    final c = j['cta'];
    if (c is Map) {
      cta = CtaContent(
        label: '${c['label'] ?? ''}',
        kind: c['kind'] as String?,
        url: c['url'] as String?,
        faq: c['faq'] is bool ? c['faq'] as bool : null,
      );
    }
    return _Localized(
      title: '${j['title'] ?? ''}',
      body: '${j['body'] ?? ''}',
      cta: cta,
    );
  }
}

class _TourStep {
  final Map<String, String> ar;
  final Map<String, String> en;
  final String? anchor;
  const _TourStep({required this.ar, required this.en, this.anchor});
}

class _ManifestAction {
  final String id;
  final String type; // popup | banner | tooltip | tour | drawer
  final _Localized arContent;
  final _Localized enContent;
  final _Localized? arContentB;
  final _Localized? enContentB;
  final String triggerKind;
  final int? triggerSeconds;
  final String? triggerEventName;
  final String? urlContains;
  final int? frequencyCap;
  final String? goalEvent;
  final String? anchorSelector;
  final List<_TourStep>? steps;

  _ManifestAction({
    required this.id,
    required this.type,
    required this.arContent,
    required this.enContent,
    this.arContentB,
    this.enContentB,
    required this.triggerKind,
    this.triggerSeconds,
    this.triggerEventName,
    this.urlContains,
    this.frequencyCap,
    this.goalEvent,
    this.anchorSelector,
    this.steps,
  });

  bool get hasB => arContentB != null || enContentB != null;

  static _ManifestAction? fromJson(Object? raw) {
    if (raw is! Map) return null;
    final id = raw['id'];
    final type = raw['type'];
    final content = raw['content'];
    final trigger = raw['trigger'];
    if (id is! String || type is! String || content is! Map || trigger is! Map) {
      return null;
    }
    final contentB = raw['contentB'];
    List<_TourStep>? steps;
    final rawSteps = raw['steps'];
    if (rawSteps is List) {
      steps = rawSteps.whereType<Map>().map((s) {
        final ar = (s['ar'] is Map) ? s['ar'] as Map : const {};
        final en = (s['en'] is Map) ? s['en'] as Map : const {};
        return _TourStep(
          ar: {'title': '${ar['title'] ?? ''}', 'body': '${ar['body'] ?? ''}'},
          en: {'title': '${en['title'] ?? ''}', 'body': '${en['body'] ?? ''}'},
          anchor: s['anchor'] as String?,
        );
      }).toList();
    }
    return _ManifestAction(
      id: id,
      type: type,
      arContent: _Localized.fromJson((content['ar'] as Map?) ?? const {}),
      enContent: _Localized.fromJson((content['en'] as Map?) ?? const {}),
      arContentB: (contentB is Map && contentB['ar'] is Map)
          ? _Localized.fromJson(contentB['ar'] as Map)
          : null,
      enContentB: (contentB is Map && contentB['en'] is Map)
          ? _Localized.fromJson(contentB['en'] as Map)
          : null,
      triggerKind: '${trigger['kind'] ?? ''}',
      triggerSeconds: trigger['seconds'] is num ? (trigger['seconds'] as num).toInt() : null,
      triggerEventName: trigger['eventName'] as String?,
      urlContains: raw['urlContains'] as String?,
      frequencyCap: raw['frequencyCap'] is num ? (raw['frequencyCap'] as num).toInt() : null,
      goalEvent: raw['goalEvent'] as String?,
      anchorSelector: raw['anchorSelector'] as String?,
      steps: steps,
    );
  }
}

const _capsKey = 'tracki_caps';

/// Stable 50/50 split per visitor (mirrors shared pickVariant). Replicates the
/// JS `(h * 31 + code) | 0` 32-bit signed overflow exactly via [toSigned].
String pickVariant(String anonId, String actionId, bool hasB) {
  if (!hasB) return 'A';
  var h = 0;
  final s = '$anonId:$actionId';
  for (var i = 0; i < s.length; i++) {
    h = (h * 31 + s.codeUnitAt(i)).toSigned(32);
  }
  return h.abs() % 2 == 0 ? 'A' : 'B';
}

/// Mobile action engine — the snippet's actions runtime re-imagined for apps:
/// screen_view plays the role of pageview, app_background plays exit_intent,
/// `event` fires on track(); struggle actions stay server-driven (Live Assist).
/// Frequency caps persist across launches in one storage-backed JSON map.
class ActionEngine {
  final String key;
  final String endpoint;
  final String Function() locale;
  final String Function() currentPath;
  final String Function() anonId;
  final EventQueue queue;
  final Transport transport;
  final KeyValueStorage storage;
  final Renderer? renderer;
  final CtaRouter cta;
  final Clock now;

  final List<_ManifestAction> _actions = [];
  final Set<String> _shownThisScreen = {};
  final Set<String> _shownThisSession = {};
  Map<String, int> _caps = {};
  bool _capsLoaded = false;
  final List<Timer> _timers = [];

  ActionEngine({
    required this.key,
    required this.endpoint,
    required this.locale,
    required this.currentPath,
    required this.anonId,
    required this.queue,
    required this.transport,
    required this.storage,
    required this.renderer,
    required this.cta,
    required this.now,
  });

  void _emit(String type, String actionId, String variant, [String? channel]) {
    final props = <String, Object?>{'action_id': actionId, 'variant': variant};
    if (channel != null) props['channel'] = channel;
    queue.enqueue(EventInput(
      type: type,
      ts: now(),
      path: currentPath(),
      props: props,
    ));
  }

  Future<void> _loadCaps() async {
    try {
      final raw = await storage.get(_capsKey);
      final parsed = raw == null ? {} : jsonDecode(raw);
      _caps = (parsed is Map)
          ? parsed.map((k, v) => MapEntry('$k', (v as num).toInt()))
          : {};
    } catch (_) {
      _caps = {};
    }
    _capsLoaded = true;
  }

  void _bumpCap(_ManifestAction a) {
    _caps[a.id] = (_caps[a.id] ?? 0) + 1;
    storage.set(_capsKey, jsonEncode(_caps)).catchError((_) {});
  }

  LocalizedContent _localized(_ManifestAction a, String variant) {
    final L = locale();
    if (variant == 'B' && a.hasB) {
      final b = L == 'en' ? (a.enContentB ?? a.arContentB) : (a.arContentB ?? a.enContentB);
      if (b != null) {
        return LocalizedContent(title: b.title, body: b.body, cta: b.cta);
      }
    }
    final pack = L == 'en' ? a.enContent : a.arContent;
    return LocalizedContent(title: pack.title, body: pack.body, cta: pack.cta);
  }

  List<TourStepContent>? _tourSteps(_ManifestAction a) {
    if (a.type != 'tour' || a.steps == null || a.steps!.isEmpty) return null;
    final L = locale();
    return a.steps!.map((s) {
      final m = L == 'en' ? s.en : s.ar;
      return TourStepContent(
        title: m['title'] ?? '',
        body: m['body'] ?? '',
        anchor: s.anchor,
      );
    }).toList();
  }

  void _maybeShow(_ManifestAction a) {
    final r = renderer;
    if (r == null) return;
    if (_shownThisScreen.contains(a.id)) return;
    if (a.urlContains != null && !currentPath().contains(a.urlContains!)) return;
    if (a.frequencyCap != null && (_caps[a.id] ?? 0) >= a.frequencyCap!) return;
    final variant = pickVariant(anonId(), a.id, a.hasB);
    final content = _localized(a, variant);
    final intent = ActionIntent(
      actionId: a.id,
      type: a.type,
      variant: variant,
      content: content,
      steps: _tourSteps(a),
      anchor: a.anchorSelector,
      activateCta: () {
        final c = content.cta;
        if (c == null) return;
        final kind = cta.activate(c);
        _emit('action_click', a.id, variant, kind.name);
      },
      dismiss: () => _emit('action_dismiss', a.id, variant),
    );
    try {
      r.show(intent);
    } catch (_) {
      // Audit slice-3 M1 parity: a failed render is NOT an impression.
      return;
    }
    _shownThisScreen.add(a.id);
    _shownThisSession.add(a.id);
    _bumpCap(a);
    _emit('action_impression', a.id, variant);
  }

  void _evaluateScreen() {
    for (final a in _actions) {
      if (a.triggerKind == 'pageview') _maybeShow(a);
    }
  }

  /// Fetch the mobile-surface manifest, then arm time-based triggers.
  Future<void> start() async {
    await _loadCaps();
    try {
      final data = await transport.get(
        '$endpoint/v1/actions?key=${Uri.encodeComponent(key)}&surface=mobile',
      );
      final list = (data is Map) ? data['actions'] : null;
      if (list is List) {
        for (final raw in list) {
          final a = _ManifestAction.fromJson(raw);
          if (a != null) _actions.add(a);
        }
      }
    } catch (_) {
      // fail-silent
    }
    _evaluateScreen();
    for (final a in _actions) {
      if (a.triggerKind == 'time_on_page') {
        final secs = a.triggerSeconds ?? 5;
        _timers.add(Timer(Duration(seconds: secs), () => _maybeShow(a)));
      }
    }
  }

  /// New screen — screen-scoped impressions reset, pageview triggers re-run.
  void onScreen() {
    _shownThisScreen.clear();
    if (_capsLoaded) _evaluateScreen();
  }

  /// App went to background — the mobile analogue of exit intent.
  void onBackground() {
    for (final a in _actions) {
      if (a.triggerKind == 'exit_intent') _maybeShow(a);
    }
  }

  /// A custom event — event triggers + goal attribution.
  void onTrack(String name) {
    for (final a in _actions) {
      if (a.triggerKind == 'event' && a.triggerEventName == name) _maybeShow(a);
      if (a.goalEvent == name && _shownThisSession.contains(a.id)) {
        _emit('action_goal', a.id, pickVariant(anonId(), a.id, a.hasB));
      }
    }
  }

  /// Test seam.
  int get loadedCount => _actions.length;
}

File: packages/mobile-core/src/actions.ts
import type { CtaRouter } from "./cta";
import type { EventQueue } from "./queue";
import type {
  ActionIntent,
  Clock,
  KeyValueStorage,
  LocalizedContent,
  Renderer,
  TourStepContent,
  Transport,
} from "./types";

// Local mirror of the mobile manifest entry (zero deps, like the snippet's).
interface Localized {
  title: string;
  body: string;
  cta?: { label: string; kind?: string; url?: string; faq?: boolean };
}
interface ManifestAction {
  id: string;
  type: "popup" | "banner" | "tooltip" | "tour" | "drawer";
  content: { ar: Localized; en: Localized };
  contentB?: { ar: Localized; en: Localized };
  trigger: { kind: string; seconds?: number; eventName?: string };
  urlContains?: string;
  frequencyCap?: number;
  goalEvent?: string;
  anchorSelector?: string;
  steps?: Array<{
    ar: { title: string; body: string };
    en: { title: string; body: string };
    anchor?: string;
  }>;
}

const CAPS_KEY = "tracki_caps";

/** Stable 50/50 split per visitor (mirrors shared pickVariant). */
export function pickVariant(anonId: string, actionId: string, hasB: boolean): "A" | "B" {
  if (!hasB) return "A";
  let h = 0;
  const s = `${anonId}:${actionId}`;
  for (let i = 0; i < s.length; i++) h = (h * 31 + s.charCodeAt(i)) | 0;
  return Math.abs(h) % 2 === 0 ? "A" : "B";
}

interface EngineDeps {
  key: string;
  endpoint: string;
  locale: () => "ar" | "en";
  currentPath: () => string;
  anonId: () => string;
  queue: EventQueue;
  transport: Transport;
  storage: KeyValueStorage;
  renderer?: Renderer;
  cta: CtaRouter;
  now: Clock;
}

/**
 * Mobile action engine — the snippet's actions runtime re-imagined for apps:
 * screen_view plays the role of pageview, app_background plays exit_intent,
 * `event` fires on track(); struggle actions stay server-driven (Live Assist).
 * Frequency caps persist across launches in one storage-backed JSON map.
 */
export function createActionEngine(deps: EngineDeps) {
  const actions: ManifestAction[] = [];
  const shownThisScreen = new Set<string>();
  const shownThisSession = new Set<string>();
  let caps: Record<string, number> = {};
  let capsLoaded = false;

  const emit = (type: string, actionId: string, variant: string, channel?: string) =>
    deps.queue.enqueue({
      type,
      ts: deps.now(),
      path: deps.currentPath(),
      props: { action_id: actionId, variant, ...(channel ? { channel } : {}) },
    });

  async function loadCaps(): Promise<void> {
    try {
      caps = JSON.parse((await deps.storage.get(CAPS_KEY)) ?? "{}") as Record<string, number>;
    } catch {
      caps = {};
    }
    capsLoaded = true;
  }

  function bumpCap(a: ManifestAction): void {
    caps[a.id] = (caps[a.id] ?? 0) + 1;
    void deps.storage.set(CAPS_KEY, JSON.stringify(caps)).catch(() => {});
  }

  function localized(a: ManifestAction, variant: "A" | "B"): LocalizedContent {
    const pack = variant === "B" && a.contentB ? a.contentB : a.content;
    return pack[deps.locale()];
  }

  function tourSteps(a: ManifestAction): TourStepContent[] | undefined {
    if (a.type !== "tour" || !a.steps?.length) return undefined;
    const L = deps.locale();
    return a.steps.map((s) => ({ ...s[L], anchor: s.anchor }));
  }

  function maybeShow(a: ManifestAction): void {
    if (!deps.renderer) return;
    if (shownThisScreen.has(a.id)) return;
    if (a.urlContains && !deps.currentPath().includes(a.urlContains)) return;
    if (a.frequencyCap && (caps[a.id] ?? 0) >= a.frequencyCap) return;
    const variant = pickVariant(deps.anonId(), a.id, !!a.contentB);
    const content = localized(a, variant);
    const intent: ActionIntent = {
      intent: "action",
      actionId: a.id,
      type: a.type,
      variant,
      content,
      steps: tourSteps(a),
      anchor: a.anchorSelector,
      activateCta: () => {
        if (!content.cta) return;
        const kind = deps.cta.activate(content.cta);
        emit("action_click", a.id, variant, kind);
      },
      dismiss: () => emit("action_dismiss", a.id, variant),
    };
    try {
      deps.renderer.show(intent);
    } catch {
      // Audit slice-3 M1 parity: a failed render is NOT an impression.
      return;
    }
    shownThisScreen.add(a.id);
    shownThisSession.add(a.id);
    bumpCap(a);
    emit("action_impression", a.id, variant);
  }

  function evaluateScreen(): void {
    for (const a of actions) if (a.trigger.kind === "pageview") maybeShow(a);
  }

  /** Fetch the mobile-surface manifest, then arm time-based triggers. */
  async function start(): Promise<void> {
    await loadCaps();
    try {
      const data = (await deps.transport.get(
        `${deps.endpoint}/v1/actions?key=${encodeURIComponent(deps.key)}&surface=mobile`,
      )) as { actions?: ManifestAction[] } | undefined;
      if (Array.isArray(data?.actions)) actions.push(...data.actions);
    } catch {
      /* fail-silent */
    }
    evaluateScreen();
    for (const a of actions) {
      if (a.trigger.kind === "time_on_page") {
        const t = setTimeout(() => maybeShow(a), (a.trigger.seconds ?? 5) * 1000);
        (t as { unref?: () => void }).unref?.();
      }
    }
  }

  return {
    start,
    /** New screen — screen-scoped impressions reset, pageview triggers re-run. */
    onScreen(): void {
      shownThisScreen.clear();
      if (capsLoaded) evaluateScreen();
    },
    /** App went to background — the mobile analogue of exit intent. */
    onBackground(): void {
      for (const a of actions) if (a.trigger.kind === "exit_intent") maybeShow(a);
    },
    /** A custom event — event triggers + goal attribution. */
    onTrack(name: string): void {
      for (const a of actions) {
        if (a.trigger.kind === "event" && a.trigger.eventName === name) maybeShow(a);
        if (a.goalEvent === name && shownThisSession.has(a.id)) {
          emit("action_goal", a.id, pickVariant(deps.anonId(), a.id, !!a.contentB));
        }
      }
    },
    /** Test seam. */
    get loadedCount(): number {
      return actions.length;
    },
  };
}

export type ActionEngine = ReturnType<typeof createActionEngine>;
